### PR TITLE
Wyujie

### PR DESCRIPTION
**Pull request by Yujie-W**
add plant hydraulics details

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,2 @@
+{
```
Comment (charleskawczynski) [inline .vscode/settings.json]: Please remove this file

File: src/Plant/characters.txt [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# Greek
+Α    α    ,
```
Comment (charleskawczynski) [inline src/Plant/characters.txt]: This file probably belongs in `docs/` as a `.md` file

File: Project.toml [hunk 3]
```diff
@@ -9,8 +9,10 @@ DelimitedFiles = "8bb1440f-4735-579b-a4ab-409b98df4dab"
 DifferentialEquations = "0c46a032-eb83-5123-abaf-570d42b7fbaa"
 DocStringExtensions = "ffbed154-4ef7-542d-bbb7-c09d3a79fcae"
 GSL = "92c85e6c-cbff-5e0c-80f7-495c94daaecd"
+HDF5 = "f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f"
 IncGammaBeta = "5087d2c0-5ad2-51de-b1ef-eaf82f0743fd"
 Interpolations = "a98d9a8b-a2ab-59e6-89dd-64a1c18fca59"
+LaTeXStrings = "b964fa9f-0449-5b57-a5c2-d3ea65f4040f"
```
Comment (charleskawczynski) [inline Project.toml]: Please remove LaTeXStrings from the deps

File: src/Plant/hydraulics/get_p_base_q_list_from_q.jl [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# this function returns the p_base from total transpiration rate
+function get_p_base_q_list_from_q(tree::StructTree, flow::Number)
```
Comment (charleskawczynski) [inline src/Plant/hydraulics/get_p_base_q_list_from_q.jl]: Can we document what equation is being solved in this routine? It looks like we should utilize some existing root-finding/numerical methods in a separate package.

File: src/Plant/hydraulics/get_struct_p_end_from_q.jl [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        dp       = flow / shell_k
+        p_end   -= dp
+    end
+
+    # compute the pressure drop along the xylem
+    for i in 1:length(any_struct.p_element)
+        p_25   = min(any_struct.p_history[i], p_end / get_relative_surface_tension(any_struct.t_element[i]))
+        k_25   = any_struct.k_element[i] * exp( -1.0 * (-p_25/any_struct.b) ^ (any_struct.c) )
+        k      = k_25 / get_relative_viscosity(any_struct.t_element[i])
+        p_end -= flow / k + 998.0 * 9.8 * any_struct.z_element[i] * 1E-6
```
Comment (charleskawczynski) [inline src/Plant/hydraulics/get_struct_p_end_from_q.jl]: Can we remove these hard-coded numbers? If `9.8` is gravity, then we should probably grab this from [CLIMAParameters](https://github.com/CliMA/CLIMAParameters.jl) using
```
g = FT(grave(param_set))
```

Which means you'll need to construct a `param_set` and pass this through.

Comment (charleskawczynski) [reply src/Plant/hydraulics/get_struct_p_end_from_q.jl]: Please refer to its use in ClimateMachine.jl

File: src/Plant/photosynthesis/get_leaf_an_ag_r_from_pi.jl [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# get a from ci
+function get_leaf_an_ag_r_from_pi(;
+                                  v25::Number = 80.0,
```
Comment (charleskawczynski) [inline src/Plant/photosynthesis/get_leaf_an_ag_r_from_pi.jl]: Please follow the convention in ClimateMachine.jl, where `FT` is used to indicate the float type.
```suggestion
                                  v25::FT = FT(80.0),
```

File: src/Plant/photosynthesis/get_leaf_an_ag_r_from_pi.jl [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+                                  j25::Number = 135.0,
+                               Γ_star::Number = 2.5,
+                                  p_i::Number = 30.0,
+                                  tem::Number = 298.15,
+                                  par::Number = 1000.0,
+                                 p_O₂::Number = 21278.25,
+                                  r25::Number = Inf,
+                                 unit::String = "K")
```
Comment (charleskawczynski) [inline src/Plant/photosynthesis/get_leaf_an_ag_r_from_pi.jl]:
```suggestion
                                 unit::String = "K") where {FT}
```
This will be needed along with my above comment

File: src/Plant/photosynthesis/get_leaf_r_from_r25.jl [hunk 8]
```diff
@@ -0,0 +1,16 @@
+# get respiration rate from r25 and leaf temperature
+function get_leaf_r_from_r25(r25::Number, tem::Number; unit::String="K")
+    # convert to degree C
+    if unit=="K" || unit=="k"
```
Comment (charleskawczynski) [inline src/Plant/photosynthesis/get_leaf_r_from_r25.jl]: Please convert all units to SI

File: src/Plant/struct/struct_tree_canopy.jl [hunk 9]
```diff
@@ -0,0 +1,78 @@
+# struct for leaf
+Base.@kwdef mutable struct StructTreeLeaf
```
Comment (charleskawczynski) [inline src/Plant/struct/struct_tree_canopy.jl]:
```suggestion
Base.@kwdef mutable struct StructTreeLeaf{FT<:AbstractFloat}
```

File: src/Plant/struct/struct_tree_canopy.jl [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    # leaf structure
+    angle_locat::Float32 = 0.0    # degree | direction from the tree trunk to the leaf, 0 for east, 270 for south, and 180 for west
```
Comment (charleskawczynski) [inline src/Plant/struct/struct_tree_canopy.jl]:
```suggestion
    angle_locat::FT = FT(0.0)    # degree | direction from the tree trunk to the leaf, 0 for east, 270 for south, and 180 for west
```
Change others alike

File: src/Plant/struct/struct_tree_canopy.jl [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    p_history::Array{Float32,1} = zeros(10)            # MPa           | a list of trunk xylem pressure history per element
+    t_element::Array{Float32,1} = ones(10) * 298.15    # K             | a list of stem temperature for each element
+    z_element::Array{Float32,1} = ones(10) * 0.0       # m             | a list of trunk height per element
+end
+
+
+
+
+# struct of canopy layer
+Base.@kwdef mutable struct StructTreeCanopyLayer
```
Comment (charleskawczynski) [inline src/Plant/struct/struct_tree_canopy.jl]:
```suggestion
Base.@kwdef mutable struct StructTreeCanopyLayer{FT<:AbstractFloat}
```

File: src/Plant/struct/struct_tree_canopy.jl [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    # canopy layer and leaf structure
+    f_layer::Float32 = 1.0      #    | fraction of LA of the layer, f_layer = 1.0 / n_layer
```
Comment (charleskawczynski) [inline src/Plant/struct/struct_tree_canopy.jl]:
```suggestion
    f_layer::FT = FT1.0)      #    | fraction of LA of the layer, f_layer = 1.0 / n_layer
```
Change others alike

File: src/Plant/tool/plot_a_pi_curve.jl [hunk 13]
```diff
@@ -0,0 +1,31 @@
+using LaTeXStrings
+using PyPlot
```
Comment (charleskawczynski) [inline src/Plant/tool/plot_a_pi_curve.jl]: Plotting utilities should be migrated into `docs/`

File: src/Plant/water/get_relative_surface_tension.jl [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    #=
+    gamma        = V^(-2/3) * k * (tem_c-tem)
+    gamma/gamm25 = (tem_c - tem) / (tem_c - 298.15)
+    the empirical values are
+    k     = 2.1E-7 J K^-1 mol^(-2/3)
+    V     = 18.0 ml/mol
+    tem_c = 647.0 K
+    =#
+    temm = 647.0
+    if unit=="C" || unit=="c"
```
Comment (charleskawczynski) [inline src/Plant/water/get_relative_surface_tension.jl]: Please convert to SI units

File: src/Plant/constants.jl [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+
+# set the default floating type, Float32 for now. Should be the same as in the RT module
+FT = Float64
+
+# more pending, may need to migrate this to CLIMAParameters in the future
+ρ_H₂O = FT(998.0) # Kg m⁻³ | @ 293.15 K, may need to calculate it from temperature in the future
+
+# load the constant gravity value
+struct EarthParameterSet <: AbstractEarthParameterSet end
+earth   = EarthParameterSet()
```
Comment (charleskawczynski) [inline src/Plant/constants.jl]:
```suggestion
const earth   = EarthParameterSet()
```

File: src/Plant/dynamic/update_tree_with_time.jl [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+# function to update the tree with time, default at a Δt=1s using Wang-2020 scheme
+function update_tree_with_time(tree::StructTree, Δt::FT=FT(1.0); scheme::String="Wang2020", updating=false) where {FT}
+    # 0. unpack necessary structs
+    @unpack branch_list = tree.branch
+    @unpack canopy_list = tree.canopy
+    @unpack root_list   = tree.roots
+
+    # 1. use the gsw from last time instant and update e and a first, because Tleaf was from the last instant
+    for canopyi::StructTreeCanopyLayer in canopy_list
```
Comment (charleskawczynski) [inline src/Plant/dynamic/update_tree_with_time.jl]:
```suggestion
    for canopyi in canopy_list
```

File: src/Plant/dynamic/update_tree_with_time.jl [hunk 17]
```diff
@@ -0,0 +1,75 @@
+using Parameters
+
+# function to update the tree with time, default at a Δt=1s using Wang-2020 scheme
+function update_tree_with_time(tree::StructTree, Δt::FT=FT(1.0); scheme::String="Wang2020", updating=false) where {FT}
```
Comment (charleskawczynski) [inline src/Plant/dynamic/update_tree_with_time.jl]: Can we add doc strings to this? [here](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Common/MoistThermodynamics/relations.jl#L73) is a good example of a doc string that follows our coding style

Comment (charleskawczynski) [inline src/Plant/dynamic/update_tree_with_time.jl]:
```suggestion
function update_tree_with_time!(tree::Tree, Δt::FT=FT(1.0); scheme::String="Wang2020", updating=false) where {FT}
```
Functions that change the contents of the arguments, by convention, append `!` to the function name.

Also, we don’t need to include `Struct` in the `Tree` type

File: src/Plant/hydraulics/get_struct_p_end_from_q.jl [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+    n   = any_struct.soil_n
+    k   = any_struct.k_rhiz
+    k_s = get_relative_surface_tension(any_struct.t_soil)
+    k_t = get_relative_viscosity(any_struct.t_soil)
+    for i in 0:9
+        if p_end<0
+            shell_t = (1.0 / (1.0 + (a*(-p_end/k_s))^n)) ^ m
+        else
+            shell_t = 1.0
+        end
```
Comment (charleskawczynski) [inline src/Plant/hydraulics/get_struct_p_end_from_q.jl:33]: Add `FT` for type stability, also, can we incorporate these hard-codes numbers into a model?

File: src/Plant/struct/struct_tree.jl [hunk 19]
```diff
@@ -0,0 +1,14 @@
+# struct for a C3 tree
+Base.@kwdef mutable struct StructTree{FT<:AbstractFloat}
+    # tree information
+    age::Int = 10          # year | age
+    ba ::FT  = FT( 0.1)    # m²   | basal area
+    ga ::FT  = FT(50.0)    # m²   | ground area
+    h  ::FT  = FT( 8.0)    # m    | tree height
+
+    # tree formation from root to leaves
+    roots ::StructTreeRoot   = StructTreeRoot{FT,5}()           # | root struct with 5 layers by default
```
Comment (charleskawczynski) [inline src/Plant/struct/struct_tree.jl]:
```suggestion
    roots ::Root   = Root{FT,5}()           # | root struct with 5 layers by default
```
Id probably suggest a shorter name here, since it’s a property of `Tree`, the `Tree` in `TreeRoot` is inferred.

File: src/Plant/water/get_saturated_vapor_pressure.jl [hunk 20]
```diff
@@ -0,0 +1,14 @@
+# this function returns the saturated vapor pressure in Pa
+function get_saturated_vapor_pressure(tem::FT) where {FT}
```
Comment (charleskawczynski) [inline src/Plant/water/get_saturated_vapor_pressure.jl:11]: We should probably use the function in MoistThermodynamics, but we’ll have to wait until it’s been pulled out of ClimateMachine.jl

Comment (charleskawczynski) [reply src/Plant/water/get_saturated_vapor_pressure.jl:11]: We can add `FT` for type stability

File: src/Plant/constants.jl [hunk 21]
```diff
@@ -1,12 +1,77 @@
...
+# pre-defined ΔE and ΔP to avoid repeated FT conversion
+const ΔEP_2 = FT(1e-2)
+const ΔEP_3 = FT(1e-3)
+const ΔEP_4 = FT(1e-4)
+const ΔEP_5 = FT(1e-5)
+const ΔEP_6 = FT(1e-6)
+const ΔEP_7 = FT(1e-7)
+
+# pre-defined number to avoid repeated FT conversion in the code
+const NUMB_0   = FT( 0.0)
```
Comment (charleskawczynski) [inline src/Plant/constants.jl]: Using `FT` in the code is fine, and it's not always needed. Please read the [conversion and promotion](https://docs.julialang.org/en/v1/manual/conversion-and-promotion/#conversion-and-promotion-1) docs to get a better idea of when they're needed. Defining module-global constants like this should be avoided.

File: src/Plant/photosynthesis/get_leaf_j.jl [hunk 22]
```diff
@@ -1,8 +1,10 @@
...
+This function calculates j from par using a quadratic function
+"""
 function get_leaf_j(jmax::FT, par::FT) where {FT}
-    a =  0.9
-    b = -0.3 * par - jmax
-    c =  0.3 * par * jmax
-    j = ( -b - sqrt(b*b-4*a*c) ) / a * 0.5
+    b = PS_J_QY * par + jmax
+    c = PS_J_QY * par * jmax
+    j = ( b - sqrt(b^NUMB_2 - NUMB_4*PS_J_CR*c) ) / PS_J_CR * NUMB_0_5
```
Comment (charleskawczynski) [inline src/Plant/photosynthesis/get_leaf_j.jl]:
```suggestion
    j = ( b - sqrt(b^2 - 4*PS_J_CR*c) ) / PS_J_CR * FT(0.5)
```
Based on [conversion and promotion](https://docs.julialang.org/en/v1/manual/conversion-and-promotion/#conversion-and-promotion-1) rules, the `2` and `4` will promote to the type `FT`, but the `0.5` must be type-casted with `FT`. This is a nice final form since it does not depend on any module-global variables, and can therefore be better optimized by the compiler.

File: src/Plant/stomata/get_marginal_gain.jl [hunk 23]
```diff
@@ -1,9 +1,13 @@
-# function to calculate marginal gain for any optimization model
-function get_marginal_gain(canopyi::StructTreeCanopyLayer, indx::Number)
-    leaf_e1   = canopyi.e_list[indx] + 1E-6
+"""
+    get_marginal_gain(canopyi, indx)
+This function calculates marginal gain for the indx th leaf in the canopyi, useful for any optimization model.
+"""
```
Comment (charleskawczynski) [inline src/Plant/stomata/get_marginal_gain.jl]:
```suggestion
"""
    get_marginal_gain(canopyi, indx)

The marginal gain, given
 - `indx` the indx'th leaf    (if this is an accurate description)
 - `canopyi` the ith canopy    (if this is an accurate description)
useful for any optimization model.
"""
```

File: src/Plant/struct/struct_tree_root.jl [hunk 24]
```diff
@@ -1,4 +1,12 @@
-# struct for root layer
+"""
+    struct RootLayer{FT}
+
+# Arguments
+- `FT`    Floating type for the RootLayer
+
+# Description
+Calling RootLayer{FT}() will create one RootLayer struct for the root system.
+"""
```
Comment (charleskawczynski) [inline src/Plant/struct/struct_tree_root.jl]:
```suggestion
"""
    struct RootLayer{FT}

One RootLayer struct for the root system.

# Fields
$(DocStringExtensions.FIELDS)
"""
```
Here, `FT` is a [type parameter](https://docs.julialang.org/en/v1/manual/types/#Parametric-Types-1). For structs, we should probably document the arguments, and this can be done nicely with `DocStringExtensions`, you can see an example [here](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Common/MoistThermodynamics/states.jl#L55-L80)

Comment (charleskawczynski) [reply src/Plant/struct/struct_tree_root.jl]: You'll need to modify the doc strings in the structs, too

File: src/Plant/struct/struct_tree_root.jl [hunk 25]
```diff
@@ -1,4 +1,12 @@
...
 Base.@kwdef mutable struct RootLayer{FT<:AbstractFloat}
     # root structure
```
Comment (charleskawczynski) [inline src/Plant/struct/struct_tree_root.jl]:
```suggestion
    "root depth `[m]`"
```

File: src/Plant/struct/struct_tree_root.jl [hunk 26]
```diff
@@ -1,4 +1,12 @@
...
     z_lo   ::FT = FT(-1.0)    # m | root depth
```
Comment (charleskawczynski) [inline src/Plant/struct/struct_tree_root.jl]:
```suggestion
    z_lo   ::FT = FT(-1.0)
    "the location for upper root `[m]`"
```

**Review by charleskawczynski — COMMENTED** [review]
I think the only remaining suggestions I have are:

 - Update doc strings based on suggestions
 - Extract out hard-coded constants, and incorporate them into model parameters
 - Rebase / squash (I can help with this if needed)

It would be nice to add tests, too, because virtually none of this code is currently tested in CI, as coverage indicates.
